third_party/gtp_ovs/ovs-gtp-patches/2.15/0023-QFI-Support-in-OVS.patch

```diff
@@ -0,0 +1,1034 @@
+ 					nla_get_be64(a), is_mask);
+ 			tun_flags |= TUNNEL_KEY;
+@@ -817,6 +820,7 @@ static int ip_tun_from_nlattr(const struct nlattr *attr,
+ 			opts_type = type;
+ 			break;
+ 		case OVS_TUNNEL_KEY_ATTR_GTPU_OPTS:
++                        OVS_NLERR(log, "OVS_TUNNEL_KEY_ATTR_GTPU_OPTS");
+ 			if (opts_type) {
+ 				OVS_NLERR(log, "Multiple metadata blocks provided");
+ 				return -EINVAL;
+@@ -830,6 +834,12 @@ static int ip_tun_from_nlattr(const struct nlattr *attr,
+ 			tun_flags |= TUNNEL_GTPU_OPT;
+ 			opts_type = type;
+ 			break;
++                case OVS_TUNNEL_KEY_ATTR_QFI:
++                        OVS_NLERR(log, "OVS_TUNNEL_KEY_ATTR_QFI");
++                        SW_FLOW_KEY_PUT(match, tun_key.qfi,
++                                        nla_get_u8(a), is_mask);
++                        tun_flags |= TUNNEL_QFI;
++                        break;
+ 
+ 		default:
+ 			OVS_NLERR(log, "Unknown IP tunnel attribute %d",
+@@ -944,9 +954,6 @@ static int __ip_tun_to_nlattr(struct sk_buff *skb,
+ 	if (output->tp_dst &&
+ 	    nla_put_be16(skb, OVS_TUNNEL_KEY_ATTR_TP_DST, output->tp_dst))
+ 		return -EMSGSIZE;
+-	if ((output->tun_flags & TUNNEL_OAM) &&
+-	    nla_put_flag(skb, OVS_TUNNEL_KEY_ATTR_OAM))
+-		return -EMSGSIZE;
+ 	if (swkey_tun_opts_len) {
+ 		if (output->tun_flags & TUNNEL_GENEVE_OPT &&
+ 		    nla_put(skb, OVS_TUNNEL_KEY_ATTR_GENEVE_OPTS,
+@@ -964,6 +971,9 @@ static int __ip_tun_to_nlattr(struct sk_buff *skb,
+ 				 swkey_tun_opts_len, tun_opts))
+ 			return -EMSGSIZE;
+ 	}
++        if (output->qfi &&
++            nla_put_u8(skb, OVS_TUNNEL_KEY_ATTR_QFI, output->qfi))
++                return -EMSGSIZE;
+ 
+ 	return 0;
+ }
+diff --git a/datapath/linux/compat/gtp.c b/datapath/linux/compat/gtp.c
+index 725db82..91d6174 100644
+--- a/datapath/linux/compat/gtp.c
++++ b/datapath/linux/compat/gtp.c
+@@ -84,6 +84,17 @@ static int check_header(struct sk_buff *skb, int len)
+ 	return 0;
+ }
+ 
++struct gtpu_ext_hdr n_hdr = {
++        .type = 0x85,
++};
++
++struct gtpu_ext_hdr_pdu_sc pdu_sc_hdr = {
++        .len = 1,
++        .pdu_type = 0x0, /* PDU_TYPE_DL_PDU_SESSION_INFORMATION */
++        .qfi = 5,
++                .next_type = 0,
++};
++
+ static int gtp_rx(struct sock *sk, struct gtp_dev *gtp, struct sk_buff *skb,
+ 			unsigned int hdrlen, u8 gtp_version,
+ 			__be64 tid, u8 flags, u8 type)
+@@ -116,15 +127,50 @@ static int gtp_rx(struct sock *sk, struct gtp_dev *gtp, struct sk_buff *skb,
+ 		if (unlikely(opts_len)) {
+ 			struct gtpu_metadata *opts = ip_tunnel_info_opts(&tun_dst->u.tun_info);
+ 			struct gtp1_header *gtp1 = (struct gtp1_header *)(skb->data + sizeof(struct udphdr));
+-
+-			opts->ver = GTP_METADATA_V1;
+-			opts->flags = gtp1->flags;
+-			opts->type = gtp1->type;
++			struct gtpu_ext_hdr *geh;
++			geh = (struct gtpu_ext_hdr *) (gtp1 + 1);
++			n_hdr.type = geh->type;
++			struct gtpu_ext_hdr_pdu_sc *pdu_sc_hd;
++			pdu_sc_hd = (struct gtpu_ext_hdr_pdu_sc *) (geh + 1);
```

Comment:
before you parse next header, you need to check if this is of PDU_SC type or not.

Reply:
Yes. Fixed it.